File: SustitucionMOAModel/Dto/OrdenDeCarga/ObtenerContratosDisponiblesResponse.cs
using SustitucionMOAModel.Enums;
using SustitucionMOAModel.Entities;
using SustitucionMOAModel.Models.DataAgro;
using SustitucionMOAModel.Models.WSMapMOA.OrdenCarga;
using System;
using System.Collections.Generic;
using System.Linq;

namespace SustitucionMOAModel.Dto.OrdenDeCarga
{
    public class ObtenerContratosDisponiblesResponse
    {
        public List<ContratoOrdenFas> Contratos { get; set; }

        public string Info { get; set; }

        public string Error { get; set; }

        public bool Logout { get; set; }
    }

    public class ContratoOrdenFas
    {
        public string NumeroContrato { get; set; }

        public MaterialDto Producto { get; set; }
        public decimal? KgDisponibles { get; set; }
        public TipoContratoFAS TipoContrato { get; set; }
        public CondicionRetiro CondicionRetiro { get; set; }

        public string DescripcionProducto
        {
            get
            {
                return !string.IsNullOrEmpty(Producto.Abreviacion) ? Producto.Abreviacion : Producto.NombreProducto;
            }
        }

        public ContratoOrdenFas(Result contratoSAP, List<Material> productosBD)
        {
            var producto = productosBD
                            .Where(p => p.CodigoSap == contratoSAP.Producto.Trim().TrimStart('0'))
                            .Select(p => new MaterialDto
                            {
                                MaterialId = p.Id,
                                Descripcion = p.Nombre,
                                CodigoSap = p.CodigoSap,
                                Abreviacion = p.Abreviacion,
                            })
                            .Single();
            NumeroContrato = contratoSAP.Contrato;
            Producto = producto;
            TipoContrato = contratoSAP.TipoContrato;
            CondicionRetiro = contratoSAP.PrecioFlete == 0 ? CondicionRetiro.RetiroEnPlanta : CondicionRetiro.PuestoEnDestino;
        }

        public ContratoOrdenFas(Entities.OrdenDeCarga orden)
        {
            NumeroContrato = orden.ContratoIngresado;
            var producto = orden.Producto;

            Producto = new Models.DataAgro.MaterialDto
            {
                MaterialId = orden.Producto_Id,
                Descripcion = producto?.Nombre,
                Abreviacion = producto?.Abreviacion
            };
        }
    }
}

File: SustitucionMOA/src/app/common/models/ordenes-de-carga/obtenerContratosDisponiblesResponse.ts
import { Formatter } from "../../formatter/Formatter";
import { Material } from "../material";

export interface ObtenerContratosDisponiblesResponse {
    Contratos: ContratoOrdenFas[];
    Info: string;
    Error: string;
    Logout: boolean;
}

export enum TipoContrato {
    Normal = 0,
    FacturaAnticipada = 1,
    CyO = 3,
}

export enum CondicionRetiro {
    PuestoEnDestino = 0,
    RetiroEnPlanta = 1
}

export class ContratoOrdenFas {
    NumeroContrato: string;
    Producto: Material;
    TipoContrato: TipoContrato;
    KgDisponibles: number;
    CondicionRetiro: CondicionRetiro;
    NombreProducto: string;
    Label: string;

    constructor(
        numeroContrato?: string,
        tipoContrato?: TipoContrato,
        producto?: Material,
        kgDisponibles?: number,
        condicionRetiro?: CondicionRetiro) {
            
        this.NumeroContrato = numeroContrato || this.NumeroContrato;
        this.KgDisponibles = kgDisponibles === null || kgDisponibles === undefined ? this.KgDisponibles : kgDisponibles;
        this.Producto = producto || this.Producto;
        this.TipoContrato = tipoContrato;
        this.CondicionRetiro = condicionRetiro;
        const prod = this.Producto.Abreviacion || this.Producto.Descripcion;
        
        const kgs = this.KgDisponibles <= 0 ? "sin kg. Disp." :
            !this.KgDisponibles ? '' :
                `${Formatter.formatNumberWithPoint(this.KgDisponibles.toString())} kg. Disp.`;

        const marcaTipoContrato = this.TipoContrato == TipoContrato.FacturaAnticipada ? ' (A)' : '';
        const marcaCondicionRetiro = this.CondicionRetiro == CondicionRetiro.PuestoEnDestino ? ' (PD)' : '';

        this.Label = `${this.NumeroContrato} - ${prod} ${kgs}${marcaTipoContrato}${marcaCondicionRetiro}`;
    }
}